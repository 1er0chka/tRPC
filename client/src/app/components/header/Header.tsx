import React, {FunctionComponent, useEffect, useState} from "react";
import styles from "./Header.module.scss";
import {Coin, Portfolio} from "@/app/service/Types";
import Service from "@/app/service/Service";
import PopularCoin from "@/app/components/header/popular-coin/PopularCoin";
import {formatNumber} from "@/app/service/Formats";
import Image from "next/image";
import ModalPortfolio from "@/app/components/modal-portfolio/ModalPartfolio";

const Header: FunctionComponent<{ coins: Coin[] }> = ({coins}) => {
    const [portfolioSum, setPortfolioSum] = useState<number>(0);
    const [portfolioDif, setPortfolioDif] = useState<number>(0);
    const [isModalVisible, setModalVisible] = useState<boolean>(false);

    useEffect(() => {
        document.addEventListener("storage", getSum);
        getSum();
        return () => document.removeEventListener("storage", getSum);
    }, []);

    const getSum = () => {
        let sum = 0;
        const portfolio = localStorage.getItem("portfolio");
        if (portfolio) {
            const data: Portfolio[] = JSON.parse(portfolio);
            data.forEach((item) => {
                sum += item.oldPrice;
            });
        }
        setPortfolioSum(sum);
    };

    useEffect(() => {
        setPortfolioDif(0);
        const getAssets = async (coin: Portfolio) => {
            Service.getAssetsById(coin.id)
                .then((data) => {
                    setPortfolioDif(
                        portfolioDif +
                        (coin.oldPrice -
                            coin.number * parseFloat(data?.priceUsd as string)),
                    );
                })
                .catch((err) => {
                    console.error(err);
                });
        };
        const portfolio = localStorage.getItem("portfolio");
        if (portfolio) {
            const data: Portfolio[] = JSON.parse(portfolio);
            data.forEach((item) => {
                getAssets(item);
            });
        }
    }, [portfolioSum]);

    return (
        <div className={styles.body}>
            {coins.length == 3 ? (
                <div className={styles.popular}>
                    <PopularCoin coin={coins[0]}/>
                    <PopularCoin coin={coins[1]}/>
                    <PopularCoin coin={coins[2]}/>
                </div>
            ) : (
                <div/>
            )}
            <div className={styles.portfolio} onClick={() => setModalVisible(true)}>
                <div className={styles.portfolioPrice}>
                    <div className={styles.price}>{formatNumber(portfolioSum)}</div>
                    {portfolioDif != 0 ? (
                        <div className={styles.difference}>
                            {formatNumber(portfolioDif)} (
                            {((portfolioDif / portfolioSum) * 100).toFixed(2)}%)
                        </div>
                    ) : (
                        <div className={styles.difference}>0 (0%)</div>
                    )}
                </div>
                <div className={styles.image}>
                    <Image
                        width={300}
                        height={300}
                        layout={"responsive"}
                        src={"/resources/images/coin.png"}
                        alt={"Portfolio"}
                    />
                </div>
            </div>
            <ModalPortfolio isVisible={isModalVisible} setVisible={setModalVisible}/>
        </div>
    );
};

export default Header;
