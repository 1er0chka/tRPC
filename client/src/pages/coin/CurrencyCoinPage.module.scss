@import "src/app/styles/fonts";

.content {
  margin: 3vw 10vw 3vw 10vw;

  @media (max-width: 1440px) {
    margin: 3vw 5vw 3vw 5vw;
  }
  @media (max-width: 1280px) {
    margin: 3vw 0 3vw 2vw;
  }
}

.info {
  display: flex;

  @media (max-width: 768px) {
    flex-direction: column-reverse;
  }
}

.infoLeft {
  width: 54vw;

  @media (max-width: 1440px) {
    width: 72vw;
  }
  @media (max-width: 768px) {
    width: 93vw;
  }
}

.back {
  @include text(500);
  text-decoration: none;
  margin-left: 1vw;
}

.infoRight {
  width: 20vw;
  margin-left: 5vw;
  margin-top: 1.7vw;

  @media (max-width: 768px) {
    width: 50vw;
  }
}

.coinPrice {
  @include title(500);
  margin-bottom: 2vw;
}

.error {
  @include title(500);
  margin-left: 1vw;
  margin-bottom: 1vw;
}

.addButton {
  @include text(500);
  margin-top: 2vw;
  background-color: rgba(255, 255, 255, 0.4);
  border: 0.1vw solid rgba(59, 39, 21, 0.3);
  border-radius: 0.1vw;
  padding: 0.2vw 0.8vw 0.2vw 0.8vw;
  width: 100%;

  &:hover {
    background-color: rgba(255, 255, 255, 0.6);
    cursor: pointer;
  }

  &:disabled {
    background-color: rgba(255, 255, 255, 0.2);
    color: rgba(59, 39, 21, 0.7);
    cursor: auto;
  }
}
