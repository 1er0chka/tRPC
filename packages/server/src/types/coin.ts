export interface Coin {
  id: string;
  name: string;
  rank: string;
  symbol: string;
  supply: string;
  maxSupply: string;
  priceUsd: string;
  changePercent24Hr: string;
  marketCapUsd: string;
}

export interface History {
  priceUsd: string;
  time: string;
  circulatingSupply: string;
  date: string;
}