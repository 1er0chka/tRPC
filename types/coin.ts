export interface Coin {
    id: string;
    name: string;
    rank: string;
    symbol: string;
    supply: string;
    maxSupply: string;
    priceUsd: string;
    changePercent24Hr: string;
    marketCapUsd: string;
}

export const defaultCoin: Coin = {
    changePercent24Hr: "",
    id: "",
    marketCapUsd: "",
    maxSupply: "",
    name: "",
    priceUsd: "",
    rank: "",
    supply: "",
    symbol: "",
};

export interface History {
    priceUsd: string;
    time: string;
    circulatingSupply: string;
    date: string;
}

export interface Portfolio {
    id: string;
    number: number;
    oldPrice: number;
}

export interface PortfolioCoin extends Portfolio {
  newPrice: number;
  difference: number;
  name: string;
  symbol: string
}

export type Sort = "rank" | "priceUsd" | "changePercent24Hr" | "marketCapUsd";
