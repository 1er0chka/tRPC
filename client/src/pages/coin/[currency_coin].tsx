import "../../app/styles/globals.css";
import React, { useContext, useEffect, useState } from "react";
import { GetStaticPaths, GetStaticProps } from "next";
import CoinPage from "@/app/components/coin-page/CoinPage";
import Header from "@/app/components/header/Header";
import Service from "@/app/service/Service";
import { useRouter } from "next/router";
import { CoinInfoContext } from "@/app/provider/CoinInfoContext";
import Loading from "@/app/components/loading/Loading";
import styles from "@/pages/coin/CurrencyCoinPage.module.scss";
import Link from "next/link";
import { Coin } from "@/app/service/Types";

export type CurrencyCoinProps = {
  slug: string;
};

const CurrencyCoinPage = ({ slug }: CurrencyCoinProps) => {
  const router = useRouter();
  const { coin, setCoin, interval, setTime, setPrice } =
    useContext(CoinInfoContext);
  const [loading, setLoading] = useState<boolean>(true);
  const [isExist, setExist] = useState<boolean>(false);
  const [popularCoins, setPopularCoins] = useState<Coin[]>([]);

  useEffect(() => {
    const getAssets = async () => {
      if (typeof router.query.currency_coin === "string") {
        Service.getAssetsById(router.query.currency_coin)
          .then((data) => {
            setCoin(data);
            setExist(true);
          })
          .catch((err) => {
            console.error(err);
            setExist(false);
            setLoading(false);
          });
      }
    };
    if (slug != undefined) {
      setLoading(true);
      getAssets();
    }
  }, [slug]);

  useEffect(() => {
    const getHistory = async () => {
      Service.getHistoryById(coin.id, interval).then((data) => {
        const x: string[] = [];
        const y: number[] = [];
        switch (interval) {
          case "m1": {
            data.map((time) => {
              x.push(
                new Date(parseInt(time.time)).getUTCHours().toString() +
                  ":" +
                  new Date(parseInt(time.time)).getUTCMinutes().toString(),
              );
              y.push(parseFloat(time.priceUsd));
            });
            break;
          }
          default: {
            data.map((time) => {
              x.push(
                new Date(parseInt(time.time)).getUTCDay().toString() +
                  "." +
                  new Date(parseInt(time.time)).getUTCMonth().toString() +
                  " " +
                  new Date(parseInt(time.time)).getUTCHours().toString() +
                  ":" +
                  new Date(parseInt(time.time)).getUTCMinutes().toString(),
              );
              y.push(parseFloat(time.priceUsd));
            });
            break;
          }
        }
        setTime(x);
        setPrice(y);
        setLoading(false);
      });
    };

    if (coin.id.length > 0) {
      getHistory();
    }
  }, [coin, interval]);

  useEffect(() => {
    setPopularCoins([]);
    const getAssets = async (id: string) => {
      Service.getAssetsById(id)
        .then((coin) => {
          setPopularCoins((prevCoins) => [...prevCoins, coin]);
        })
        .catch((err) => {
          console.error(err);
          setPopularCoins([]);
        });
    };
    getAssets("bitcoin");
    getAssets("ethereum");
    getAssets("tether");
  }, []);

  return (
    <div>
      <Header coins={popularCoins} />
      {loading ? (
        <Loading />
      ) : (
        <div>
          {isExist ? (
            <CoinPage />
          ) : (
            <div className={styles.content}>
              <div className={styles.error}>Page doesn`t exist</div>
              <Link className={styles.back} href={"/"}>
                {" "}
                ‚Üê Back to table
              </Link>
            </div>
          )}
        </div>
      )}
    </div>
  );
};
export default CurrencyCoinPage;

export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: [],
    fallback: true,
  };
};

export const getStaticProps: GetStaticProps<CurrencyCoinProps> = async (
  context,
) => {
  const slug = context.params!.currency_coin as string;
  return {
    props: {
      slug,
    },
  };
};
