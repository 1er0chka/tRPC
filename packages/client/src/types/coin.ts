import {Coin} from "../../../../types/coin";

export const defaultCoin: Coin = {
    changePercent24Hr: "",
    id: "",
    marketCapUsd: "",
    maxSupply: "",
    name: "",
    priceUsd: "",
    rank: "",
    supply: "",
    symbol: "",
};

export interface Portfolio {
    id: string;
    number: number;
    oldPrice: number;
}

export interface PortfolioCoin extends Portfolio {
  newPrice: number;
  difference: number;
  name: string;
  symbol: string
}

export type Sort = "rank" | "priceUsd" | "changePercent24Hr" | "marketCapUsd";
