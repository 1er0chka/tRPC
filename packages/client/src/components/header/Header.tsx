import React, {FunctionComponent, useState} from 'react';
import styles from './Header.module.scss'
import {Coin, PortfolioCoin} from "../../../../../types/coin";
import HeaderCoinCard from "./header-coin-card/HeaderCoinCard";
import ModalPortfolio from "../modal-portfolio/ModalPortfolio";
import UserPortfolio from "./portfolio/UserPortfolio";

interface IHeaderProps {
    coins: Coin[]
    portfolioCoins: PortfolioCoin[]
}

const Header: FunctionComponent<IHeaderProps> = ({coins, portfolioCoins}) => {
    const [isModalVisible, setModalVisible] = useState<boolean>(false);

    return (
        <div className={styles.body}>
            {coins.length == 3 ?
                <div className={styles.popularCoins}>
                    {coins.map((coin) => <HeaderCoinCard coin={coin}/>)}
                </div>
                :
                <div/>
            }
            <UserPortfolio onClickAction={() => setModalVisible(true)}/>
            <ModalPortfolio isVisible={isModalVisible} setVisible={setModalVisible} portfolioCoins={portfolioCoins}/>
        </div>
    );
};

export default Header;