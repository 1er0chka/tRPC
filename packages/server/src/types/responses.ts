import {Coin, History} from "../../../../types/coin";


export interface ICoinsResponse {
    data: Coin[];
}

export interface ICoinResponse {
    data: Coin;
}

export interface IHistoryResponse {
    data: History[];
}
