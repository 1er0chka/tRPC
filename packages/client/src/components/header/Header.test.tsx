import {render, screen} from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import {Coin} from "../../../../../types/coin";
import Header from "./Header";

const coins: Coin[] = [{
    changePercent24Hr: "-0.8101417214350335",
    id: "bitcoin",
    marketCapUsd: "119179791817.6740161068269075",
    maxSupply: "21000000.0000000000000000",
    name: "Bitcoin",
    priceUsd: "6931.5058555666618359",
    rank: "1",
    supply: "17193925.0000000000000000",
    symbol: "BTC"
}, {
    changePercent24Hr: "-2.3610937754922363",
    id: "ethereum",
    marketCapUsd: "260178024694.6417172413108697",
    maxSupply: "null",
    name: "Ethereum",
    priceUsd: "2164.5806540967609070",
    rank: "2",
    supply: "120197888.76991934",
    symbol: "ETH"
}, {
    changePercent24Hr: "-0.811417214350335",
    id: "test",
    marketCapUsd: "11179791817.6740161068269075",
    maxSupply: "2000000.0000000000000000",
    name: "test",
    priceUsd: "631.5058555666618359",
    rank: "1",
    supply: "17193925.0000000000000000",
    symbol: "TST"
}]

const refresh = jest.fn()
const portfolioSum: number = 1299.99
const portfolioDif: number = 188.32

describe('Header Test', () => {
    test('Snapshot test', () => {
        render(
            <Header coins={coins} portfolioRefresh={refresh} portfolioSum={portfolioSum} portfolioDif={portfolioDif}/>
        )
        const header = screen.queryByTestId('header')
        expect(header).toMatchSnapshot()
    })

    test('Empty test', () => {
        render(
            <Header coins={[]} portfolioRefresh={refresh} portfolioSum={portfolioSum} portfolioDif={0.0}/>
        )
        const header = screen.queryByTestId('header')
        expect(header).toMatchSnapshot()
    })

    test('User Portfolio Call Test', async () => {
        render(
            <Header coins={[]} portfolioRefresh={refresh} portfolioSum={portfolioSum} portfolioDif={portfolioDif}/>
        )
        const button = screen.queryByTestId('portfolio')
        if (button) {
            await userEvent.click(button)
        }
    })
})