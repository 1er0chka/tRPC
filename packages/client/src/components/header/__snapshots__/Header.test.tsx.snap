// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Test Empty test 1`] = `
<div
  class="body"
  data-testid="header"
>
  <div />
  <div
    class="body"
    data-testid="portfolio"
  >
    <div
      class="portfolioSum"
      data-testid="portfolio-sum"
    >
      <div
        class="summa"
      >
        $1,299.99
      </div>
      <div
        class="difference"
        data-testid="portfolio-difference"
      >
        0 (0%)
      </div>
    </div>
    <img
      alt=""
      class="image"
      src="/resources/images/coin.png"
    />
  </div>
</div>
`;

exports[`Header Test Snapshot test 1`] = `
<div
  class="body"
  data-testid="header"
>
  <div
    class="popularCoins"
  >
    <div
      class="body"
      data-testid="header-coin-card"
    >
      <div
        class="coinInfo"
      >
        <img
          alt=""
          class="coinImage"
          src="https://assets.coincap.io/assets/icons/btc@2x.png"
        />
        <div
          class="coinName"
          data-testid="coin-name"
        >
          Bitcoin
        </div>
        <div
          class="coinSymbol"
          data-testid="coin-symbol"
        >
          BTC
        </div>
      </div>
      <div
        class="coinPrice"
        data-testid="coin-price"
      >
        $6,931.51
      </div>
    </div>
    <div
      class="body"
      data-testid="header-coin-card"
    >
      <div
        class="coinInfo"
      >
        <img
          alt=""
          class="coinImage"
          src="https://assets.coincap.io/assets/icons/eth@2x.png"
        />
        <div
          class="coinName"
          data-testid="coin-name"
        >
          Ethereum
        </div>
        <div
          class="coinSymbol"
          data-testid="coin-symbol"
        >
          ETH
        </div>
      </div>
      <div
        class="coinPrice"
        data-testid="coin-price"
      >
        $2,164.58
      </div>
    </div>
    <div
      class="body"
      data-testid="header-coin-card"
    >
      <div
        class="coinInfo"
      >
        <img
          alt=""
          class="coinImage"
          src="https://assets.coincap.io/assets/icons/tst@2x.png"
        />
        <div
          class="coinName"
          data-testid="coin-name"
        >
          test
        </div>
        <div
          class="coinSymbol"
          data-testid="coin-symbol"
        >
          TST
        </div>
      </div>
      <div
        class="coinPrice"
        data-testid="coin-price"
      >
        $631.51
      </div>
    </div>
  </div>
  <div
    class="body"
    data-testid="portfolio"
  >
    <div
      class="portfolioSum"
      data-testid="portfolio-sum"
    >
      <div
        class="summa"
      >
        $1,299.99
      </div>
      <div
        class="difference"
        data-testid="portfolio-difference"
      >
        $188.32
         (
        14.49
        %)
      </div>
    </div>
    <img
      alt=""
      class="image"
      src="/resources/images/coin.png"
    />
  </div>
</div>
`;
